src/core/ddsi/src/ddsi_typewrap.c

```diff
@@ -443,36 +475,388 @@ static void set_member_detail (struct xt_member_detail *dst, const DDS_XTypes_Co
+  dds_return_t ret = DDS_RETCODE_OK;
+  uint32_t cnt = t->_u.union_type.members.length;
+  if (cnt == 0)
+  {
+    GVTRACE ("union has no members\n");
+    return DDS_RETCODE_BAD_PARAMETER;
+  }
+  DDS_XTypes_MemberId *ids = ddsrt_malloc (cnt * sizeof (*ids));
+  for (uint32_t n = 0; n < cnt; n++)
+    ids[n] = t->_u.union_type.members.seq[n].id;
+  qsort (ids, cnt, sizeof (*ids), xt_member_id_cmp);
+  for (uint32_t n = 0; n < cnt - 1; n++)
+  {
+    if (ids[n] == ids[n + 1])
+    {
+      GVTRACE ("duplicate member id %"PRIu32" in union\n", ids[n]);
+      ret = DDS_RETCODE_BAD_PARAMETER;
+      goto err;
+    }
+  }
+
+err:
+  ddsrt_free (ids);
+  return ret;
+}
+
+static int xt_enum_value_cmp (const void *va, const void *vb)
+{
+  const int32_t *m1 = va, *m2 = vb;
+  return (*m1 == *m2) ? 0 : (*m1 < *m2) ? -1 : 1;
+}
+
+static dds_return_t xt_valid_enum_values (struct ddsi_domaingv *gv, const struct xt_type *t)
+{
+  assert (!ddsi_xt_is_unresolved (t) && t->_d == DDS_XTypes_TK_ENUM);
+  dds_return_t ret = DDS_RETCODE_OK;
+  uint32_t cnt = t->_u.enum_type.literals.length;
+  int32_t *values = ddsrt_malloc (cnt * sizeof (*values));
+  for (uint32_t n = 0; n < cnt; n++)
+    values[n] = t->_u.enum_type.literals.seq[n].value;
+  qsort (values, cnt, sizeof (*values), xt_enum_value_cmp);
+  for (uint32_t n = 0; n < cnt - 1; n++)
+  {
+    if (values[n] == values[n + 1])
+    {
+      GVTRACE ("duplicate enum value %"PRIi32"\n", values[n]);
+      ret = DDS_RETCODE_BAD_PARAMETER;
+      goto err;
+    }
+  }
+
+err:
+  ddsrt_free (values);
+  return ret;
+}
+
+static int xt_bitmask_position_cmp (const void *va, const void *vb)
+{
+  const uint16_t *m1 = va, *m2 = vb;
+  return (*m1 == *m2) ? 0 : (*m1 < *m2) ? -1 : 1;
+}
+
+static dds_return_t xt_valid_bitmask_positions (struct ddsi_domaingv *gv, const struct xt_type *t)
+{
+  assert (!ddsi_xt_is_unresolved (t) && t->_d == DDS_XTypes_TK_BITMASK);
+  dds_return_t ret = DDS_RETCODE_OK;
+  uint32_t cnt = t->_u.bitmask.bitflags.length;
+  uint16_t *positions = ddsrt_malloc (cnt * sizeof (*positions));
+  for (uint32_t n = 0; n < cnt; n++)
+    positions[n] = t->_u.bitmask.bitflags.seq[n].position;
+  qsort (positions, cnt, sizeof (*positions), xt_bitmask_position_cmp);
+  for (uint32_t n = 0; n < cnt - 1; n++)
+  {
+    if (positions[n] == positions[n + 1])
+    {
+      GVTRACE ("duplicate bitmask position %"PRIu16"\n", positions[n]);
+      ret = DDS_RETCODE_BAD_PARAMETER;
+      goto err;
+    }
+  }
+
+err:
+  ddsrt_free (positions);
+  return ret;
+}
+
+#define F DDS_XTypes_IS_FINAL
+#define A DDS_XTypes_IS_APPENDABLE
+#define M DDS_XTypes_IS_MUTABLE
+#define N DDS_XTypes_IS_NESTED
+#define H DDS_XTypes_IS_AUTOID_HASH
+static dds_return_t xt_valid_type_flags (struct ddsi_domaingv *gv, uint16_t flags, uint8_t type_kind)
+{
+  dds_return_t ret = DDS_RETCODE_OK;
+  switch (type_kind)
+  {
+    case DDS_XTypes_TK_STRUCTURE:
+    case DDS_XTypes_TK_UNION:
+      if (flags & ~(F|A|M|N|H))
```

Comment:
In most (places of the DDS) specs, unused flags are to be ignored to allow for future extensibility. It may be different in this case.